AK/Lexer.cpp

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright (c) 2020, Benoit Lormeau <blormeau@outlook.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Assertions.h>
+#include <AK/Lexer.h>
+
+namespace AK {
+
+Lexer::Lexer(const StringView& input)
+    : m_input(input)
+{
+}
+
+Lexer::~Lexer()
+{
+}
+
+// Tells whether the parser's index has reached input's end
+bool Lexer::eof() const
+{
+    return m_index >= m_input.length();
+}
+
+// Returns the current character at the parser index, plus `offset` if specified
+char Lexer::peek(size_t offset) const
+{
+    return (m_index + offset < m_input.length()) ? m_input[m_index + offset] : '\0';
+}
+
+// Tests the next character in the input
+bool Lexer::next_is(char expected) const
+{
+    return peek() == expected;
+}
+
+// Tests if the `expected` string comes next in the input
+bool Lexer::next_is(const char* expected) const
+{
+    for (size_t i = 0; peek(i) && expected[i]; ++i)
+        if (peek(i) != expected[i])
+            return false;
+    return true;
```

Comment:
You're right about this. I was thinking about an efficient way to do it in a single for loop, but I think a simple check at the beginning will suffice. What do you think?
```suggestion
    if (tell_remaining() < __builtin_strlen(expected))
        return false;

    for (size_t i = 0; expected[i] != '\0'; ++i)
        if (peek(i) != expected[i])
            return false;
    return true;
```

Reply:
Yes, that could work too ! I'm not sure if `__builtin_strlen()` would necessarily be more efficient than putting it in the loop though, since computing the strlen would also need to walk the string. But it would make it clearer as well ^^

Reply:
My bad, for some reason I thought it wouldn't work without the little check. I did some testing and it does work :)